armory/instrument/instrument.py

```diff
@@ -660,18 +680,37 @@ class ResultsWriter(Writer):
     Write results to dictionary
     """
 
-    def __init__(self, sink=None):
+    def __init__(self, sink=None, max_record_size=None):
         """
         sink is a callable that takes the output results dict as input
             if sink is None, call get_output after close to get results dict
+        max_record_size - if not None, the maximum size in bytes for a single record
+            Records exceeding that size log a warning and are dropped
         """
         super().__init__()
         self.sink = sink
         self.records = []
         self.output = None
+        if max_record_size is not None:
+            if max_record_size < -1:
+                raise ValueError(f"max_record_size {max_record_size} cannot be < -1")
+            if json_utils is None:
+                raise ValueError("Cannot import numpy. Set max_record_size to None")
+            max_record_size = int(max_record_size)
+        self.max_record_size = max_record_size
 
     def _write(self, name, batch, result):
-        self.records.append((name, batch, result))
+        record = (name, batch, result)
+        if self.max_record_size is not None:
+            try:
+                json_utils.check_size(record, self.max_record_size)
+            except ValueError:
+                log.warning(
+                    f"record ({name}, {batch}, ...) size > "
```

Comment:
might not be super clear that the integer in the middle is `batch` when it's logged:

```
2022-05-10 14:54:24 13s WARNING  armory.instrument.instrument:_write:708 record (x_preprocessed, 9, ...) size > max_record_size 1048576 Dropping.

```

maybe consider `f"batch={batch}"` ?

Reply:
Updated.